secio/src/dh_compat/mod.rs

```diff
@@ -17,6 +17,56 @@ pub enum KeyAgreement {
     X25519,
 }
 
+#[cfg(all(ossl110, unix))]
+pub use openssl_impl::*;
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(not(unix))]
+pub use ring_impl::*;
+#[cfg(all(not(ossl110), unix))]
+pub use ring_openssl_unix::*;
+
+#[cfg(all(not(ossl110), unix))]
+mod ring_openssl_unix {
+    use super::KeyAgreement;
+    use super::{openssl_impl, ring_impl};
+    use crate::error::SecioError;
+
+    pub enum EphemeralPrivateKey {
+        Openssl(openssl_impl::EphemeralPrivateKey),
+        Ring(ring_impl::EphemeralPrivateKey),
+    }
+
+    pub fn generate_agreement(
+        algorithm: KeyAgreement,
+    ) -> Result<(EphemeralPrivateKey, Vec<u8>), SecioError> {
+        match algorithm {
+            KeyAgreement::EcdhP256 | KeyAgreement::EcdhP384 => {
+                openssl_impl::generate_agreement(algorithm)
+                    .map(|(private, pubkey)| (EphemeralPrivateKey::Openssl(private), pubkey))
+            }
+            KeyAgreement::X25519 => ring_impl::generate_agreement(algorithm)
+                .map(|(private, pubkey)| (EphemeralPrivateKey::Ring(private), pubkey)),
+        }
+    }
+
+    pub fn agree(
```

Comment:
is it better to change the match arms to `EphemeralPrivateKey`? which is easier to understand the code and removes the `unreachable` block.
```suggestion
    pub fn agree(
        algorithm: KeyAgreement,
        my_private_key: EphemeralPrivateKey,
        other_public_key: &[u8],
    ) -> Result<Vec<u8>, SecioError> {
        match my_private_key {
            EphemeralPrivateKey::Openssl(private_key) => {
                openssl_impl::agree(algorithm, private, other_public_key)
            }
            EphemeralPrivateKey::Ring(private_key) => {
                ring_impl::agree(algorithm, private, other_public_key)
            }
        }
    }
```

btw, is it possible to remove the `algorithm` param from the `openssl_impl::agree` fn since we have the algorithm information in the private key struct already.

Reply:
Since we need to use different implementations for different environments, the interface must be consistent with the worst implementation, and the ring implementation needs to display the specified algorithm, so the interface cannot be changed

Reply:
match arms already modified

Reply:
I don't mean modifying the interface, I mean modifying the internal `openssl_impl::agree`

```
        match my_private_key {
            EphemeralPrivateKey::Openssl(private_key) => {
                openssl_impl::agree(private_key, other_public_key)
            }
            EphemeralPrivateKey::Ring(private_key) => {
                ring_impl::agree(algorithm, private_key, other_public_key)
            }
        }
```

Reply:
```
#[cfg(all(ossl110, unix))]
pub use openssl_impl::*;
```
it also pub use when unix and assl110, so it's interface can't change

Reply:
I did not understand the issue, the `pub use` of openssl impl is added in this PR, change the internal impl fn interface should not be a breaking change issue, could you tell me in more detail what the specific problem is?

Reply:
now `dh_compat` export two function: 

```rust
    pub fn agree(
        algorithm: KeyAgreement,
        my_private_key: EphemeralPrivateKey,
        other_public_key: &[u8],
    ) -> Result<Vec<u8>, SecioError>
    
        pub fn generate_agreement(
        algorithm: KeyAgreement,
    ) -> Result<(EphemeralPrivateKey, Vec<u8>), SecioError>
```

Other modules use these two function signatures as a convention. Regardless of the internal changes, it is necessary to ensure that the two function interfaces are unchanged so that the implementation is compatible with the outside.

- `#[cfg(all(ossl110, unix))]` only use  openssl, which requires a consistent interface, so it cannot be modified
- `#[cfg(not(unix))]` only use ring
- `#[cfg(all(not(ossl110), unix))]` only use ring on x25519

In these three cases, the external interface must be the same

Reply:
I see, thanks for the detailed explanation